File: clients/macos/OpenFlowMac/BrowserHistory.swift
import Foundation
import SQLite3
import OpenFlowKit

/// Reads the domains you actually visit out of a browser's history, so the
/// vocabulary can be seeded with real sites instead of waiting for each one to
/// be transcribed wrong first.
///
/// This is the most sensitive thing OpenFlow touches, so the handling is
/// deliberately narrow:
///
/// - **Read-only, and never in place.** The database is copied to a temp file
///   and the copy is opened; the browser holds the original open and a stray
///   write would corrupt it. The copy is deleted before this returns.
/// - **Only hosts leave this function.** Paths, query strings, titles and
///   timestamps are dropped at the source -- the URL column is read but never
///   returned, so nothing downstream can leak a full URL into a file or a log.
/// - **Nothing happens without being asked.** There is no scan on launch; the
///   user presses a button, sees the list, and chooses to write it.
enum BrowserHistory {
    struct Source: Identifiable, Hashable {
        let id: String
        let name: String
        /// Bundle id, so a seeded list lands in the right vocabulary section.
        let bundleID: String
        let path: URL
    }

    /// Browsers with a readable history on this Mac.
    ///
    /// Safari is deliberately absent: its history lives under Full Disk Access,
    /// so asking would fail for most people with a permissions dialog we
    /// cannot resolve from here.
    static func available() -> [Source] {
        let fm = FileManager.default
        let home = fm.homeDirectoryForCurrentUser
        var found: [Source] = []

        let firefox = home.appending(path: "Library/Application Support/Firefox/Profiles")
        if let profiles = try? fm.contentsOfDirectory(at: firefox,
                                                      includingPropertiesForKeys: nil) {
            // Newest profile wins when there are several; "default-release" is
            // the usual one but is not guaranteed to exist.
            let places = profiles
                .map { $0.appending(path: "places.sqlite") }
                .filter { fm.fileExists(atPath: $0.path) }
                .sorted { modified($0) > modified($1) }
            if let p = places.first {
                found.append(Source(id: "firefox", name: "Firefox",
                                    bundleID: "org.mozilla.firefox", path: p))
            }
        }

        for (id, name, bundle, dir) in [
            ("chrome", "Chrome", "com.google.chrome",
             "Library/Application Support/Google/Chrome"),
            ("edge", "Edge", "com.microsoft.edgemac",
             "Library/Application Support/Microsoft Edge"),
            ("brave", "Brave", "com.brave.browser",
             "Library/Application Support/BraveSoftware/Brave-Browser"),
        ] {
            let history = home.appending(path: "\(dir)/Default/History")
            if fm.fileExists(atPath: history.path) {
                found.append(Source(id: id, name: name, bundleID: bundle, path: history))
            }
        }
        return found
    }

    private static func modified(_ url: URL) -> Date {
        (try? url.resourceValues(forKeys: [.contentModificationDateKey])
            .contentModificationDate) ?? .distantPast
    }

    enum Failure: LocalizedError {
        case unreadable(String)
        var errorDescription: String? {
            switch self {
            case .unreadable(let why): return why
            }
        }
    }

    /// Most-visited domains, highest first. Hosts only -- see the note above.
    static func domains(from source: Source, limit: Int = 400) throws
        -> [(domain: String, count: Int)] {
        let staging = FileManager.default.temporaryDirectory
            .appending(path: "of-history-\(UUID().uuidString)")
        try? FileManager.default.createDirectory(at: staging, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: staging) }

        // The write-ahead log holds recent visits; without it a browser that is
        // currently running looks weeks out of date.
        let copy = staging.appending(path: "history.sqlite")
        do {
            try FileManager.default.copyItem(at: source.path, to: copy)
            for suffix in ["-wal", "-shm"] {
                let side = URL(fileURLWithPath: source.path.path + suffix)
                guard FileManager.default.fileExists(atPath: side.path) else { continue }
                try? FileManager.default.copyItem(
                    at: side, to: URL(fileURLWithPath: copy.path + suffix))
            }
        } catch {
            throw Failure.unreadable(
                "Could not read \(source.name)'s history: \(error.localizedDescription)")
        }

        var db: OpaquePointer?
        guard sqlite3_open_v2(copy.path, &db, SQLITE_OPEN_READONLY, nil) == SQLITE_OK else {
            sqlite3_close(db)
            throw Failure.unreadable("Could not open \(source.name)'s history database.")
        }
        defer { sqlite3_close(db) }

        // Firefox and the Chromium browsers name these differently; try both
        // rather than branching on which browser we think this is.
        let queries = [
            "SELECT url, visit_count FROM moz_places WHERE visit_count > 0 ORDER BY visit_count DESC LIMIT ?;",
            "SELECT url, visit_count FROM urls WHERE visit_count > 0 ORDER BY visit_count DESC LIMIT ?;",
        ]

        for sql in queries {
            var st: OpaquePointer?
            guard sqlite3_prepare_v2(db, sql, -1, &st, nil) == SQLITE_OK else {
                sqlite3_finalize(st)
                continue                      // wrong schema; try the other
            }
            defer { sqlite3_finalize(st) }
            sqlite3_bind_int(st, 1, Int32(limit))

            var visits: [(url: String, count: Int)] = []
            while sqlite3_step(st) == SQLITE_ROW {
                guard let c = sqlite3_column_text(st, 0) else { continue }
                visits.append((String(cString: c), Int(sqlite3_column_int(st, 1))))
            }
            // Hosts are extracted here, so the URLs go out of scope with this
            // loop and never reach a caller.
            return Domains.rank(visits)
        }

        throw Failure.unreadable(
            "\(source.name)'s history is not in a layout OpenFlow understands.")
    }
}

File: clients/windows/src/win/browser_history.rs
//! Reads the domains you actually visit out of a browser's history, so the
//! vocabulary can be seeded with real sites instead of waiting for each one to
//! be transcribed wrong first.
//!
//! This is the most sensitive thing OpenFlow touches, so the handling is
//! deliberately narrow:
//!
//! - **Read-only, and never in place.** The database is copied to a temporary
//!   file and the copy is opened; the browser holds the original open and a
//!   stray write would corrupt it. The copy is deleted before this returns.
//! - **Only hosts leave this function.** Paths, query strings, titles and
//!   timestamps are dropped at the source -- the URL column is read but never
//!   returned, so nothing downstream can leak a full URL into a file or a log.
//! - **Nothing happens without being asked.** There is no scan on launch; the
//!   user presses a button, sees the list, and chooses to write it.

use std::path::PathBuf;

use rusqlite::{Connection, OpenFlags};

use crate::kit::domains;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Source {
    pub id: String,
    pub name: String,
    /// Executable name, so a seeded list lands in the right vocabulary section.
    pub app_id: String,
    pub path: PathBuf,
}

/// Browsers with a readable history on this machine.
pub fn available() -> Vec<Source> {
    let mut found = Vec::new();
    let local = std::env::var("LOCALAPPDATA").map(PathBuf::from).ok();
    let roaming = std::env::var("APPDATA").map(PathBuf::from).ok();

    if let Some(local) = &local {
        for (id, name, exe, dir) in [
            ("chrome", "Chrome", "chrome.exe", "Google/Chrome/User Data"),
            ("edge", "Edge", "msedge.exe", "Microsoft/Edge/User Data"),
            (
                "brave",
                "Brave",
                "brave.exe",
                "BraveSoftware/Brave-Browser/User Data",
            ),
            ("vivaldi", "Vivaldi", "vivaldi.exe", "Vivaldi/User Data"),
        ] {
            let history = local.join(dir).join("Default").join("History");
            if history.exists() {
                found.push(Source {
                    id: id.into(),
                    name: name.into(),
                    app_id: exe.into(),
                    path: history,
                });
            }
        }
    }

    if let Some(roaming) = &roaming {
        let profiles = roaming.join("Mozilla/Firefox/Profiles");
        if let Ok(entries) = std::fs::read_dir(&profiles) {
            // Newest profile wins when there are several; "default-release" is
            // the usual one but is not guaranteed to exist.
            let mut places: Vec<(std::time::SystemTime, PathBuf)> = entries
                .flatten()
                .map(|e| e.path().join("places.sqlite"))
                .filter(|p| p.exists())
                .filter_map(|p| {
                    let modified = std::fs::metadata(&p).ok()?.modified().ok()?;
                    Some((modified, p))
                })
                .collect();
            places.sort_by(|a, b| b.0.cmp(&a.0));
            if let Some((_, path)) = places.into_iter().next() {
                found.push(Source {
                    id: "firefox".into(),
                    name: "Firefox".into(),
                    app_id: "firefox.exe".into(),
                    path,
                });
            }
        }
    }

    found
}

/// Names the temporary copy so a leftover is recognisable, and so the tests can
/// assert none survives.
const STAGING_PREFIX: &str = "of-history-";

/// Most-visited domains, highest first. Hosts only -- see the note above.
pub fn domains_from(source: &Source, limit: i64) -> Result<Vec<(String, i64)>, String> {
    // Unique per call, not per browser: two reads in flight at once would
    // otherwise share a directory and delete each other's copy on the way out.
    static SEQUENCE: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(0);
    let staging = std::env::temp_dir().join(format!(
        "{}{}-{}",
        STAGING_PREFIX,
        std::process::id(),
        SEQUENCE.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
    ));
    let _ = std::fs::create_dir_all(&staging);
    let copy = staging.join("history.sqlite");

    let cleanup = || {
        let _ = std::fs::remove_dir_all(&staging);
    };

    // The write-ahead log holds recent visits; without it a browser that is
    // currently running looks weeks out of date.
    if let Err(e) = std::fs::copy(&source.path, &copy) {
        cleanup();
        return Err(format!("Could not read {}'s history: {e}", source.name));
    }
    for suffix in ["-wal", "-shm"] {
        let side = PathBuf::from(format!("{}{suffix}", source.path.display()));
        if side.exists() {
            let _ = std::fs::copy(&side, PathBuf::from(format!("{}{suffix}", copy.display())));
        }
    }

    let result = read_visits(&copy, limit);
    cleanup();

    match result {
        Some(visits) => Ok(domains::rank(&visits)),
        None => Err(format!(
            "{}'s history is not in a layout OpenFlow understands.",
            source.name
        )),
    }
}

fn read_visits(path: &std::path::Path, limit: i64) -> Option<Vec<(String, i64)>> {
    let db = Connection::open_with_flags(path, OpenFlags::SQLITE_OPEN_READ_ONLY).ok()?;

    // Firefox and the Chromium browsers name these differently; try both rather
    // than branching on which browser we think this is.
    for sql in [
        "SELECT url, visit_count FROM moz_places WHERE visit_count > 0 ORDER BY visit_count DESC LIMIT ?1;",
        "SELECT url, visit_count FROM urls WHERE visit_count > 0 ORDER BY visit_count DESC LIMIT ?1;",
    ] {
        let Ok(mut stmt) = db.prepare(sql) else {
            continue; // wrong schema; try the other
        };
        let rows = stmt.query_map([limit], |row| {
            Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?))
        });
        if let Ok(rows) = rows {
            // Hosts are extracted by the caller, so the URLs go out of scope
            // with this function and never reach a UI or a file.
            return Some(rows.flatten().collect());
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    /// These tests share one temporary directory and one process id, and one of
    /// them asserts that no staging copy is left behind. Run in parallel they
    /// would see each other's copies mid-read, so they take turns.
    static SERIAL: std::sync::Mutex<()> = std::sync::Mutex::new(());

    fn serial() -> std::sync::MutexGuard<'static, ()> {
        SERIAL.lock().unwrap_or_else(|e| e.into_inner())
    }

    fn chromium_fixture(path: &std::path::Path) {
        let db = Connection::open(path).unwrap();
        db.execute_batch(
            "CREATE TABLE urls(id INTEGER PRIMARY KEY, url TEXT, title TEXT, visit_count INTEGER);
             INSERT INTO urls(url,title,visit_count) VALUES
               ('https://www.doordash.com/store/1?secret=xyz','Order',9),
               ('https://news.ycombinator.com/item?id=1','HN',4),
               ('http://localhost:3000/admin','local',7),
               ('https://doordash.com/other','Order',2);",
        )
        .unwrap();
    }

    #[test]
    fn only_hosts_come_out_and_they_are_ranked_by_visits() {
        let _serial = serial();
        let dir = std::env::temp_dir().join(format!("of-bh-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("History");
        chromium_fixture(&path);

        let source = Source {
            id: "test".into(),
            name: "Test".into(),
            app_id: "chrome.exe".into(),
            path,
        };
        let domains = domains_from(&source, 100).unwrap();
        assert_eq!(
            domains,
            vec![
                ("doordash.com".to_string(), 11),
                ("news.ycombinator.com".to_string(), 4),
            ]
        );
        // The paths and query strings are gone, not merely unused.
        assert!(domains.iter().all(|(d, _)| !d.contains('/') && !d.contains("secret")));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn an_unrecognised_schema_says_so_rather_than_returning_nothing() {
        let _serial = serial();
        let dir = std::env::temp_dir().join(format!("of-bh-bad-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("History");
        Connection::open(&path)
            .unwrap()
            .execute_batch("CREATE TABLE nonsense(a INTEGER);")
            .unwrap();

        let source = Source {
            id: "test".into(),
            name: "Test".into(),
            app_id: "chrome.exe".into(),
            path,
        };
        assert!(domains_from(&source, 100).is_err());
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn a_missing_history_file_is_an_error_not_a_crash() {
        let _serial = serial();
        let source = Source {
            id: "test".into(),
            name: "Test".into(),
            app_id: "chrome.exe".into(),
            path: PathBuf::from("C:/nope/History"),
        };
        assert!(domains_from(&source, 100).is_err());
    }

    /// The original must never be opened for writing, and the copy must be
    /// gone by the time this returns.
    #[test]
    fn the_browsers_own_database_is_left_alone() {
        let _serial = serial();
        let dir = std::env::temp_dir().join(format!("of-bh-ro-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("History");
        chromium_fixture(&path);
        let before = std::fs::metadata(&path).unwrap().len();

        let source = Source {
            id: "test".into(),
            name: "Test".into(),
            app_id: "chrome.exe".into(),
            path: path.clone(),
        };
        domains_from(&source, 100).unwrap();

        assert_eq!(std::fs::metadata(&path).unwrap().len(), before);

        let mine = format!("{STAGING_PREFIX}{}-", std::process::id());
        let leftovers = std::fs::read_dir(std::env::temp_dir())
            .unwrap()
            .flatten()
            .filter(|e| e.file_name().to_string_lossy().starts_with(&mine))
            .count();
        assert_eq!(leftovers, 0, "the copy must not outlive the read");
        let _ = std::fs::remove_dir_all(&dir);
    }
}
